### PR TITLE
Added handling for OpenSSL "xts duplicated keys" error.

### LINKED ISSUE #5998: openssl aes "xts duplicated keys" error is reported as "Unknown OpenSSL Error"
Recent versions of openssl [explicitly reject duplicated AES-XTS keys](https://github.com/openssl/openssl/blob/c39352/crypto/evp/e_aes.c#L3119). Cryptography openssl backend reports this as an Unknown OpenSSL Error.

* Versions of Python, ``cryptography``, ``cffi``, ``pip``, and ``setuptools``
  you're using
  * Python 3.8.5
  * cryptopraphy main branch (f08a7de65)
  * cffi==1.14.5
  * pip 21.0.1
* How you installed ``cryptography``
  * pip install -e .
* Clear steps for reproducing your bug

```
>>> from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
>>> from cryptography.hazmat.backends import default_backend
>>> badkey = b'\xEE'*32
>>> cipher = Cipher(algorithms.AES(badkey), modes.XTS(b'\x00'*16), backend=default_backend())
>>> cipher.encryptor().update(b'\x00'*16)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/angus/tmp/cryptography/src/cryptography/hazmat/primitives/ciphers/base.py", line 109, in encryptor
    ctx = self._backend.create_symmetric_encryption_ctx(
  File "/home/angus/tmp/cryptography/src/cryptography/hazmat/backends/openssl/backend.py", line 460, in create_symmetric_encryption_ctx
    return _CipherContext(self, cipher, mode, _CipherContext._ENCRYPT)
  File "/home/angus/tmp/cryptography/src/cryptography/hazmat/backends/openssl/ciphers.py", line 115, in __init__
    self._backend.openssl_assert(res != 0)
  File "/home/angus/tmp/cryptography/src/cryptography/hazmat/backends/openssl/backend.py", line 210, in openssl_assert
    return binding._openssl_assert(self._lib, ok, errors=errors)
  File "/home/angus/tmp/cryptography/src/cryptography/hazmat/bindings/openssl/binding.py", line 80, in _openssl_assert
    raise InternalError(
cryptography.exceptions.InternalError: Unknown OpenSSL error. This error is commonly encountered when another library is not cleaning up the OpenSSL error stack. If you are using cryptography with another library that uses OpenSSL try disabling it before reporting a bug. Otherwise please file an issue at https://github.com/pyca/cryptography/issues with information on how to reproduce this. ([_OpenSSLErrorWithText(code=101511351, lib=6, func=207, reason=183, reason_text=b'error:060CF0B7:digital envelope routines:aesni_xts_init_key:xts duplicated keys')])
```

We can probably submit a patch to convert this into a more user-friendly error. If that sounds good, I was wondering how you prefer to do this. For example a couple of ideas:

1. Python-side check that throws something like a ValueError when parsing the key+mode. (This makes the assumption that cryptography is OK to always reject vulnerable duplicate keys regardless of what the backend does, so I wanted to check that's OK and not going to break anything.)
2. Check the openssl error in hazmat/backends/openssl/ciphers.py and raise something specific there based on `EVP_R_XTS_DUPLICATED_KEYS` error.

Context: This was submitted by one of our users https://github.com/espressif/esptool/issues/621

Thanks very much for all the great work on this library, we really appreciate being able to rely on it!

### PR DESCRIPTION
**Pull request by projectgus**
Closes #5998

This error value was added pre-OpenSSL 1.1.1d here:
https://github.com/openssl/openssl/commit/2a5f63c9a61be7582620c4b5da202bb3fd7e4138

and refined to only cover encryption shortly after:
https://github.com/openssl/openssl/commit/58ae5a47da1e4843b0cd1846eb297b341d0e7201
EDIT: in openssl 3.0.0 it still will cover both encryption and decryption if in FIPS mode

As far as I can tell, there is no similar error value in LibreSSL.

I'm not very familiar with CFFI or binding Python to a complex shared library, so hopefully the version checks are OK.

### CODE REVIEW

**Comment by alex** [comment]
Build appears to be failing.

**Comment by projectgus** [comment]
> Build appears to be failing.

Sorry, I messed up the conditional inclusion for `EVP_R_XTS_DUPLICATED_KEYS`. I think it's correct now (if this is wrong also then I'll find a way to test it offline.)

File: tests/hazmat/primitives/test_aes.py [hunk 1]
```diff
@@ -60,6 +60,20 @@ def test_xts_too_short(self):
         with pytest.raises(ValueError):
             enc.update(b"0" * 15)
 
+    @pytest.mark.supported(
+        only_if=lambda backend:
+        (
+            backend._lib.CRYPTOGRAPHY_OPENSSL_111D_OR_GREATER,
+        ),
```
Comment (alex) [inline tests/hazmat/primitives/test_aes.py:67]: This check isn't correct: You're returning a tuple from the callback, which always evaluates to truthy.

Comment (projectgus) [reply tests/hazmat/primitives/test_aes.py:67]: derp, thanks!

Comment (projectgus) [reply tests/hazmat/primitives/test_aes.py]: I've kept the parens, which were added to keep line length under flake8 limit - but removed the offending comma.

Comment (alex) [reply tests/hazmat/primitives/test_aes.py]: You can (and likely will need to, for CI to pass) run `black`, which will deal with all these formatting issues.

File: src/cryptography/hazmat/backends/openssl/ciphers.py [hunk 2]
```diff
@@ -112,7 +112,20 @@ def __init__(self, backend, cipher, mode, operation):
...
             operation,
         )
-        self._backend.openssl_assert(res != 0)
+
+        # Check for XTS mode duplicate keys error
+        errors = self._backend._consume_errors()
+        lib = self._backend._lib
+        if (
+                res == 0
+                and lib.CRYPTOGRAPHY_OPENSSL_111D_OR_GREATER
```
Comment (projectgus) [inline src/cryptography/hazmat/backends/openssl/ciphers.py:121]: If you prefer this test to be `Cryptography_HAS_EVP_R_XTS_DUPLICATED_KEYS`, to future proof this against the same error code being added to LibreSSL, then let me know and I'll change the logic around.

File: tests/hazmat/primitives/test_aes.py (hunk 1 above)
Comment (projectgus) [reply tests/hazmat/primitives/test_aes.py]: Seems I also missed that the `only_if` is only evaluated is only evaluated if the test function has a `backend` argument.

Comment (reaperhulk) [reply tests/hazmat/primitives/test_aes.py]: Yeah, that's a very non-obvious requirement so not a surprise you missed it initially.

**Comment by alex** [comment]
```
./tests/hazmat/primitives/test_aes.py:74:13: F841 local variable 'enc' is assigned to but never used
```

**Comment by projectgus** [comment]
Hmm, so in `py39 openssl 3.0.0-alpha17` the `xts duplicated keys` error was returned but wasn't detected as such:

> 2021-06-04T01:56:15.2964007Z [1m[31mE           cryptography.exceptions.InternalError: Unknown OpenSSL error. This error is commonly encountered when another library is not cleaning up the OpenSSL error stack. If you are using cryptography with another library that uses OpenSSL try disabling it before reporting a bug. Otherwise please file an issue at https://github.com/pyca/cryptography/issues with information on how to reproduce this. ([_OpenSSLErrorWithText(code=478150805, lib=57, func=0, reason=149, reason_text=b'error:1C800095:Provider routines::xts duplicated keys')])[0m

I'll get back to this over the next week some time and figure it out, sorry for bouncing off your CI so much.

I'm guessing if I put up a testing PR against my own fork then I can trigger CI for these configs without bothering you folks to manually trigger it for me each time?

**Comment by alex** [comment]
Not sure how github actions behaves on PRs.

FWIW, a "workaround" might be that if you send some tiny PR (find a typo fix or something :P) we can merge that and then CI will run automatically for your PRs without our intervention.

**Comment by projectgus** [comment]
>  CI / py39 openssl 3.0.0-alpha17 (pull_request) Cancelled after 20m — py39 openssl 3.0.0-alpha17

Hmm, this job last took 7 minutes on main branch so the timeout may be related to my changes. Although I'm not sure I follow how...

**Comment by alex** [comment]
Looks like a random timeout, I've restarted the job.

**Comment by alex** [comment]
Ok I have no earthly idea why this job is hanging. Can you try rebasing or somethig?

**Comment by projectgus** [comment]
Huh...

The only code change I can think of that might increase overall runtime is calling `self._backend._consume_errors()` on every cipher context now, instead of only during failures. But this looks like it should be a pretty minimal function call if there are no errors, tripling the total test runtime seems unlikely...

I'll try to set up a local environment with openssl 3 and reproduce this.

**Comment by alex** [comment]
If I were you, I'd just try rebasing. I cannot possibly imagine what in this PR would cause a hang, so I'm prepared to call this "gremlins" if rebasing fixes it.

**Comment by projectgus** [comment]
@alex This should work now, root cause was embarrassingly simple. And TIL if you generate enough IndexErrors, pytest uses a lot of resources... :)

Tests have passed locally on both openssl 1.1.1f and 3.0.0alpha17, so I think this is finally ready.

**Review by alex — APPROVED** [review]

**Pull request merged**